### DISCUSSION TITLE
How to Grid with different column widths?

### DISCUSSION THREAD

**Discussion by adsick** [emilk/egui#161] [Q&A]
I'm writing an app which has a lot to do with tables, but I'm new to Rust and Egui so it's unclear how to do nice looking tables here.

Now I have something like this: 

```
fn display_table_grid(t: &Table, ui: &mut egui::Ui) {

    egui::Grid::new("Grid")
        .min_col_width(180.0)
        .striped(true)
        .show(ui, |ui| {
            for colname in &t.colnames {
                
                // if colname == &"surname" {
                    
                //     ui.set_width(5.0);
                // } 
                ui.label(format!("{}", colname));
            }
            ui.end_row();

            ui.end_row();
            for row in &t.rowstrings {
                for entry in row {
                    //ui.set_width_range(100.0..=500.0);
                    ui.selectable_label(false, format!("{}", entry));
                }
                ui.end_row();
            }
        });
}
```

I need every row to look like a single line by making narrow columns narrower and wide wider, but it starts to wrap contents of selectable labels where there is some amount of text and the final table may look like a mess...

**Comment by adsick** [comment]
![image](https://user-images.githubusercontent.com/24528839/107115644-c0805e80-6876-11eb-9881-e64b8b499806.png)
see how redundant 'id' column gets when I use ```.min_col_width(180.0)```?

and here what it looks like without ```.min_col_width(180.0)```
![image](https://user-images.githubusercontent.com/24528839/107115715-35539880-6877-11eb-84e6-c6103c775576.png)
(the app window is fullscreen, but Grid does not take all the horizontal space available and wraps text which looks kinda uneven and ugly)

**Reply by emilk** [reply]
Nice screenshots! Time to put on some Pink Floyd.

I've improved the text wrapping logic for egui 0.9.0 (probably to be released tomorrow). 
I've also added a `Grid::max_col_width` so you can control when to wrap text in grids. If not called the text will not wrap.

PS: Roger Waters was the vocalist on lots of songs :)

**Reply by adsick** [reply]
Ok, as I know Grid is relatively new feature now. 
as a temporary fix I did this in my code:
```rust
fn display_table_grid(t: &mut Table, ui: &mut egui::Ui) {
    egui::Grid::new("Grid").striped(true).show(ui, |ui| {
        for colname in &t.colnames {
            if ui.selectable_label(false, format!("{}", colname)).clicked { //logic for sorting
                if t.order_by != Some(colname.clone()) {
                    t.order_by = Some(colname.clone())
                } else {
                    t.order_dir = !t.order_dir
                }
                t.old = true;
            }
        }
        ui.end_row();
        //gap here
        ui.end_row();
        for row in &t.rowstrings {
            for entry in row {
                egui::Label::new(entry).multiline(false).ui(ui); //fixed wrong wrapping LOOK AT ME
            }
            ui.end_row();
        }
    });
}
```
and it looks like this:
![image](https://user-images.githubusercontent.com/24528839/107144160-e7f02d80-6941-11eb-899b-dc40c9c3c377.png)
basically does the job, but does not use all the space available.
also seems like that won't work with selectable labels.

**Reply by emilk** [reply]
> `.multiline(false)`

The good news is this call won't be needed in egui 0.9.0 (widgets in grids will have text wrapping off by default). It will also work with `SelectableLabel`s.
